File: 46-permutations/permutations.java
class Solution {
    void swap(List<Integer> s,int i,int j){
        int temp=s.get(i);
        s.set(i,s.get(j));
        s.set(j,temp);
    }
    void check(int i,int[] nums,List<Integer> s,List<List<Integer>> ans){
        if(i==nums.length){
            ans.add(new ArrayList<>(s));
            return ;
        }

        for(int j=i;j<nums.length;j++){
            swap(s,i,j);
            check(i+1,nums,s,ans);
            swap(s,i,j);
        }
    }
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> ans=new ArrayList<>();
        List<Integer> s=new ArrayList<>();
        for(int i=0;i<nums.length;i++) s.add(nums[i]);
        check(0,nums,s,ans);
        return ans;
    }
}

File: 46-permutations/permutations.cpp
class Solution {
private:

void swap(vector<int>& nums, int i, int j){
    int temp = nums[i];
    nums[i] = nums[j];
    nums[j] = temp;
}

void check(int ind, vector<int>& nums, vector<vector<int>>& ans, int n){
    if(ind == n - 1){
        ans.push_back(nums);
        return ;
    }

    for(int i = ind; i < n; i ++){
        swap(nums, i, ind);
        check(ind + 1, nums, ans, n);
        swap(nums, i, ind);
    }

    return ;
}


public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> ans;
        int n = nums.size();
        check(0, nums, ans, n);
        return ans;     
    }
};